asm-5: Rework of the 5th chapter (#47)
0xAX/asm: content/asm_5.md

## Single-line NASM macros

NASM provides support for the two form of macros:

- `single-line`
- `multi-line`

All the single-line macro must start from `%define` directive. The form of this directive is following:

```assembly
%define macro_name(parameter) value
```

This is very similar to an usual C-like macro that we may define using the `#define` directive. For example, using this directive we can create the following single-line macros:

```assembly
;; Define a symbolic name for the location of argv (CLI arguments) on the stack
;; In x86-64 System V ABI, the stack layout at program start is:
;; [rsp]         → argc
;; [rsp + 8]     → argv[0]
;; [rsp + 16]    → argv[1]
;; ...

;; Pointer to the number of command line arguments
define argc [rsp]

;; Pointer to argv[0] (program name)
%define argv0 qword [rsp + 8]

;;Pointer to argv[1] (first actual CLI arg)
%define argv1 qword [rsp + 16]
```

Now we can use our definition in the code:

```assembly
;; Store the number of command line arguments in the rax register
mov rax, argc
;; Store the pointer to the first command like argument in the rsi register
mov rsi, argv1
```

> If we use the `%define` directive to define a macro, the macro code is evaluated when the macro is used. NASM provides additional directive - `%xdefine` which has the same form as the `%define` directive but it is evaluated immediately at the point of definition.

Author: Yes... expanded would be more correct

        msg_2   db      "Message 2"
        ;; ASCII code of the new line symbol ('\n').
        newline db      0xA

;; Definition of the .text section.
section .text
        ;; Reference to the entry point of our program.
        global _start

;; Entry point of the program.
_start:
        ;; Print the first message with the length 9.
        PRINT msg_1, 9
        ;; Print new line message with the length 1.
        PRINT newline, 1
        ;; Print the first message with the length 9.
        PRINT msg_2, 9
        ;; Print new line message with the length 1.
        PRINT newline, 1
        ;; Exit from the program. The 0 status code is success.
        EXIT 0
```

Now let's try to go through the macros definitions and try to understand how they work. The macro definition starts from the `%macro` directive and definition of the name of the macro and number of input parameters. Our `PRINT` macro will expect two arguments:

- The reference to the string that is going to be printed
- The length of this string

The `EXIT` macro expects only one input argument - exit code. 

There is an additional syntax - `n-*` which says that a macro accepts at least `n` arguments.

> In the macros body we just initialize the registers according to the [ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) to call the [system calls](https://en.wikipedia.org/wiki/System_call). The only one difference that you may note is that instead of direct values, we use the input parameters in the macro. The input parameters starts from the `%` symbol followed by the number of the parameter.

Author: Yes, will add

```bash
~$ nasm -f elf64 -o test.o test.asm && ld -o test test.o && ./test
Message 1
```

Although if we will pass `DEBUG` to the program or define it, we will see the both messages are printed:

```bash
~$ nasm -DDEBUG -f elf64 -o test.o test.asm && ld -o test test.o && ./test
Message 1
Message 2
```

## Useful standard macros

Besides the self-written macros, NASM provides rich set of already pre-defined macros and directives. In this section we will see some them.

### %include directive

We can include another assembly file into our file and jump to the labels defined in it, or call procedures defined there. The usage is pretty trivial, just specify the name of the assembly file you want to include:

```assembly
%include "helpers.asm"
```

### %assign directive

The single-line macro also can be defined with the special directive - `%assign`. The difference from the `%define` directive, is that using the `%assign` directive you can define a macro without parameters that will expand to a numeric value.

For example:

```assembly
%assign increment_i_var i + 1
```

### %defstr directive

> The `%defstr` directive exists to define a macro without parameters that will expand to a quoted string.

Author: without is correct here

For example:

```assembly
%defstr hello_world_msg "Hello world!"
```

### %! directive

Sometimes it can be useful to get a value of an environment variable. For this, NASM provides the special `%!` directive that can be used to get a value from the given environment variable. For example:

```assembly
%defstr HOME %!HOME
```

### User defined error directives

NASM provides the following three directives to specify user-defined errors or warnings:

- `%error msg`
- `%fatal msg`
- `%warning msg`

> In a case of the first two the execution of the program will be interrupted if one of the will be met.

Author: Yes, I will add information

        field-2-name: resb 1
endstruc
```

After we have defined our structure, we can make so-called "instance" of it. Let's take a look at the following code reworked from the previous example:

```assembly
;; Definition of the PRINT macro.
%macro PRINT 2
        ;; Specify the number of the system call (1 is `sys_write`).
    	mov rax, 1
        ;; Set the first argument of `sys_write` to 1 (`stdout`).
        mov rdi, 1
        ;; Set the second argument of `sys_write` to the reference of the string to print.
        ;; The reference will be stored in the first argument of the macro.
        mov rsi, %1
        ;; Set the third argument of `sys_write` to the length of the string to print.
        ;; The reference will be stored in the second argument of the macro.
        mov rdx, %2
        ;; Call the `sys_write` system call.
        syscall
%endmacro

;; Definition of the EXIT program
%macro EXIT 1
        ;; Specify the number of the system call (60 is `sys_exit`).
        mov rax, 60
        ;; Set the first argument of `sys_exit` to the first argument of the macro.
        mov rdi, %1
        ;; Call the `sys_exit` system call.
        syscall
%endmacro

> ;; Define person structure

Author: Yes, I like it.

```

After we have defined our macros, we can use them in our code. For example:

```assembly
;; Definition of the .data section.
section .data
        ;; The first message to print.
        msg_1   db      "Message 1"
        ;; The second message to print.
        msg_2   db      "Message 2"
        ;; ASCII code of the new line symbol ('\n').
        newline db      0xA

;; Definition of the .text section.
section .text
        ;; Reference to the entry point of our program.
        global _start

;; Entry point of the program.
_start:
        ;; Print the first message with the length 9.
        PRINT msg_1, 9
        ;; Print new line message with the length 1.
        PRINT newline, 1
        ;; Print the second message with the length 9.
        PRINT msg_2, 9
        ;; Print new line message with the length 1.
        PRINT newline, 1
        ;; Exit from the program. The 0 status code is success.
        EXIT 0
```

Now let's go through the macro definitions and see how they work. A macro definition starts from the `%macro` directive, followed by the macro’s name and the number of input parameters it expects. For example, our `PRINT` macro expects two arguments:

> - The reference to the string that is going to be printed

Reviewer: ```suggestion
- The reference to the string that will be printed
```

Reviewer: ... or `that we want to print`

- The length of this string

The `EXIT` macro expects only one input argument - exit code. 

There is also an additional syntax (`n-*`) that indicates a macro accepts at least `n` arguments.

In the macro's body, we initialize the registers according to the [ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) to make [system calls](https://en.wikipedia.org/wiki/System_call). The only difference you may notice is that instead of using direct values, we use input parameters. These parameters start with the `%` symbol, followed by the number of the parameter (`%1`, `%2`, etc).

> Another syntax ability in the multi-line NASM macros is the possibility to define labels inside a macro. In this case, the label name should be prefixed with `%%`. For example:

Reviewer: Should we mention why anyone would need to define labels? Or is it rather obvious?

- `ifmacro/endif` - check the existence of the given macro name
- `if/elif/else/endif` - check the numeric expression
- `ifenv/elifenv/elifnenv/endif` - check if an environment variable exists
- And others, like `ifempty`, `iftoken`, and `ifid`.

For example, we can slightly change the previous example and use `%ifdef DEBUG`:

```assembly
;; Entry point of the program.
_start:
        ;; Print the first message with the length 9.
        PRINT msg_1, 9
        ;; Print new line message with the length 1.
        PRINT newline, 1
%ifdef DEBUG
        ;; Print the second message with the length 9.
        PRINT msg_2, 9
        ;; Print new line message with the length 1.
        PRINT newline, 1
%endif
        ;; Exit from the program. The 0 status code is success.
        EXIT 0
```

In this code snippet, the second message will be printed only if `DEBUG` is defined. If we run the code as is, we will see that only the first message is printed:

```bash
~$ nasm -f elf64 -o test.o test.asm && ld -o test test.o && ./test
Message 1
```

However, if we pass `DEBUG` to the program or define it, we will see that both messages are printed:

```bash

> ~$ nasm -DDEBUG -f elf64 -o test.o test.asm && ld -o test test.o && ./test

Reviewer: ```suggestion
~$ nasm -DEBUG -f elf64 -o test.o test.asm && ld -o test test.o && ./test
```

### %include directive

We can include another assembly file into our file and jump to the labels defined in it, or call procedures defined there. The usage is pretty trivial - just specify the name of the assembly file you want to include:

```assembly
%include "helpers.asm"
```

### %assign directive

You can also define a single-line macro with the special directive - `%assign`. The difference from the `%define` directive is that, by using the `%assign` directive, you can define a macro without parameters that will expand to a numeric value.

For example:

```assembly
%assign increment_i_var i + 1
```

### %defstr directive

The `%defstr` directive exists to define a macro without parameters that will expand to a quoted string.

For example:

```assembly
%defstr hello_world_msg "Hello world!"
```

### %! directive

Sometimes it can be useful to get the value from the given environment variable. For this, NASM provides the special `%!` directive. For example:

```assembly
%defstr HOME %!HOME
```

### User-defined error directives

NASM provides the following directives to specify user-defined errors or warnings:

- `%error msg`
- `%fatal msg`
- `%warning msg`

> For the first two directives, the program will print the message given as an argument of the directive and the execution will interrupted. In a case of the `%warning` directive, only the message will be printed.

Reviewer: ```suggestion
For the first two directives, the program prints the message given as an argument, and the program's execution is interrupted. In the case of the `%warning` directive, only the message is printed.
```

Let's take a look at the definition of the macro `REPX` defined in the [x86inc.asm](https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/x86/x86inc.asm) source code file:

```assembly
;; Repeats an instruction/operation for multiple arguments.
;; Example usage: "REPX {psrlw x, 8}, m0, m1, m2, m3"
%macro REPX 2-* ; operation, args
    %xdefine %%f(x) %1
    %rep %0 - 1
        %rotate 1
        %%f(%1)
    %endrep
%endmacro
```

As you may see, this macro contains comments that help to understand from the beginning what it should do. Let's try to understand how this macro does its job.

> First of all, the definition of the macro is a little bit new to us. The `2-*` notation means that the macro accepts at least 2 input parameters. The macro will use the first argument as an instruction and the rest of the arguments to which this instruction will be applied.

Reviewer: ```suggestion
First of all, the definition of the macro is a little bit new to us. The `2-*` notation means that the macro accepts at least 2 input parameters. The macro uses the first argument as an instruction and applies it to the remaining arguments.
```

The first line in the macro body defines the local macro named `%%f`. This macro is expanded to the first argument of the `REPX` macro. Because the `xdefine` directive is used instead of `%define`, it captures the value of the first argument of the `REPX` macro as is. For example, `%%f` will be bound to `{psrlw x, 8}`, if the macro is used like this:

```assembly
REPX {psrlw x, 8}, m0, m1, m2, m3
```

> The macro name is prefixed with `%%` to define a uniquely scoped macro with the name which scoped only to this invocation of the `REPX` macro to not clash with other macros named `f` elsewhere.

Reviewer: ```suggestion
The macro name is prefixed with `%%` to define a uniquely scoped macro, with the name limited to this invocation of the `REPX` macro. This prevents clashes with other macros named `f` elsewhere.
```

The next instruction in this macro is the `%rep` directive. In our case, it repeats the given body `%0 - 1` times. The `%0` here is the number of the arguments given to the `REPX` macro. Since the first argument of the macro is an instruction that we need to repeat for the given arguments, we skip one repetition by doing `%0 - 1`.

The loop in this macro consists of two lines of code:

- The first line contains the `%rotate` directive, which rotates the arguments by 1 position at each iteration of the loop. It moves the second argument to the first position, the third argument to the second position, and so on.

> - The second line has the invocation of the %%f macro, which was previously defined to expand into the first argument of the REPX macro. It also replaces x in the first argument of REPX with the parameter given to the %%f macro. By passing %1 to %%f, the macro expands to the original first argument of the REPX macro, replacing the placeholder x with each of the subsequent arguments of the REPX. On each iteration of the loop, x is substituted with the next argument of the REPX macro starting from the second because of the previous rotate definition  — so the first iteration uses the second argument, the second iteration uses the third, and so on. This is repeated until all arguments are processed.

Reviewer: ```suggestion
- The second line invokes the `%%f` macro, which was previously defined to expand into the first argument of the REPX macro. It also replaces `x` in the first argument of REPX with the parameter given to the `%%f` macro. By passing `%1` to `%%f`, the macro expands to the original first argument of the REPX macro, replacing the placeholder `x` with each of the subsequent arguments of REPX. On each iteration of the loop, `x` is substituted with the next argument of the REPX macro starting from the second because of the previous rotate definition — so the first iteration uses the second argument, the second iteration uses the third, and so on. This is repeated until all arguments are processed.
```

> - The second line has the call of the `%%f` macro. Remember that this macro is bound to the first original instruction of the `REPX` macro. By passing the `%1` to the `%%f` macro it expands to the original first argument of the `REPX` macro with the `x` replaced with the first argument of the same macro but after we rotated them. So at the first iteration of the loop it will be expanded to the first argument of the `REPX` macro with `x` replaced with the second argument of the same macro. At the second iteration of the loop it will be expanded to the first argument of the `REPX` macro with `x` replaced with the third argument of the same macro. This will be repeated for the all arguments of the `REPX` macro.

Reviewer: ```suggestion
- The second line calls the `%%f` macro, which was previously defined to represent the first argument of the `REPX` macro. By passing `%1` to `%%f`, the macro expands that first argument, replacing the placeholder `x` with each of the subsequent arguments to `REPX`, one at a time. On each iteration of the loop, `x` is substituted with the next argument — so the first iteration uses the second argument, the second iteration uses the third, and so on. This is repeated until all arguments are processed.
```

So, when we use a macro with such arguments:

```assembly
REPX {psrlw x, 8}, m0, m1, m2, m3
```

It will be expanded to:

```assembly
pmulhuw m0, m7
pmulhuw m1, m7
pmulhuw m2, m7
pmulhuw m3, m7
```

> We have just tried to understand real-world assembly code!

Reviewer: ```suggestion
We’ve just explored a real-world assembly code! 🎉 🎉 🎉 
```
